### PR TITLE
Define a `grad_method` and `grad_recipe` for `CompositeOp`

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
Gradient methods are still determined using the `grad_method` property of `Operation`s. `CompositeOp` does not inherit from `Operation` and thus has not `grad_method` defined. However, it may occur within a circuit and is being queued.
This makes PennyLane try to determine the `grad_method` of the `CompositeOp`, which is not defined.

**Description of the Change:**
- Copy the `grad_method` property from `Operation` to `CompositeOp` (@albi3ro Do you think this may get us into trouble?)
- Define a dynamically computed `grad_recipe` property for `CompositeOp` that takes the number of parameters into account.
- Make `Operation.grad_method` compatible with objects that do not have an attribute `parameter_frequencies`, returning `"F"` like for the default `parameter_frequencies` property of `Operation` itself.

**Benefits:**
`CompositeOp` can be differentiated via finite differences.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

**Review by Qottmann — COMMENTED** [review]
Awesome, thanks @dwierichs !

How about an integration test from the original problem?
```
import pennylane as qml
import pennylane.numpy as pnp

dev = qml.device("default.qubit", wires=3, shots=1000)

@qml.qnode(dev)
def circuit(theta):
    qml.prod(qml.Z(0), qml.RX(theta, 1))
    return qml.expval(qml.Z(1))

par = pnp.array(1.23, requires_grad=True)
circuit(par), qml.grad(circuit)(par)
```

Note that the gradient here yields `-200000.00000000017` so there is likely still something wrong?

**Comment by dwierichs** [comment]
I think this is a problem with the `h` keyword argument of `finite_diff`, which defaults to `1e-7`, which is not suitable for `shots=1000` :grimacing: 

```python
import pennylane as qml
import pennylane.numpy as pnp

dev = qml.device("default.qubit", wires=3, shots=1000)

@qml.qnode(dev, diff_method="finite-diff")
def circuit(theta):
    qml.RX(theta, 1)
    return qml.expval(qml.Z(1))

par = pnp.array(1.23, requires_grad=True)
print(circuit(par), qml.grad(circuit)(par))
>>> 0.332 -600000.0
```

Changing the default value to `h=0.1` gives for your example with the `CompositeOp`
```python
import pennylane as qml
import pennylane.numpy as pnp

dev = qml.device("default.qubit", wires=3, shots=10000)

@qml.qnode(dev)
def circuit(theta):
    qml.prod(qml.Z(0), qml.RX(theta, 1))
    return qml.expval(qml.Z(1))

par = pnp.array(1.23, requires_grad=True)
print(circuit(par), qml.grad(circuit)(par))
>>> 0.332 -0.96
print(circuit(par), qml.grad(circuit)(par, shots=None))
>>> 0.37 -0.9424888019316975
```

So I think we should modify the default value from `h=1e-7` to something much larger, or do so only if the tape has `shots!=None`, but it is an issue unrelated to this fix :grimacing:

**Comment by Qottmann** [comment]
I think this isnt just a problem for finite diff though 😬 

```
import pennylane as qml
import pennylane.numpy as pnp

dev = qml.device("default.qubit", wires=3, shots=1000)

@qml.qnode(dev, diff_method="parameter-shift")
def circuit(theta):
    qml.prod(qml.I(0), qml.RX(theta, 1))
    # qml.RX(theta, 1)
    return qml.expval(qml.Z(1))

par = pnp.array(1.23, requires_grad=True)
circuit(par), qml.grad(circuit)(par)
import pennylane as qml
import pennylane.numpy as pnp

dev = qml.device("default.qubit", wires=3, shots=1000)

@qml.qnode(dev, diff_method="parameter-shift")
def circuit(theta):
    qml.prod(qml.I(0), qml.RX(theta, 1))
    # qml.RX(theta, 1)
    return qml.expval(qml.Z(1))

par = pnp.array(1.23, requires_grad=True)
>>> circuit(par), qml.grad(circuit)(par)
(0.374, -100000.00000000009)
```

File: pennylane/ops/op_math/composite.py [hunk 1]
```diff
@@ -129,8 +129,18 @@ def num_wires(self):
...
+        of the constituents."""
         return sum(op.num_params for op in self)
 
+    @property
+    def grad_recipe(self):
+        """The gradient recipe of the composite operator. By defauly, the recipe is
+        set to None (no parameters) or ([None]*num_params) (parameters)."""
+        if (num_params := self.num_params) == 0:
+            return None
+        return [None] * num_params
```
Comment (albi3ro) [inline pennylane/ops/op_math/composite.py:142]: What about something like:
```
recipes = []
for op in self:
    if op.grad_recipe is not None:
        recipes.extend(op.grad_recipe)
return recipes
```

Or something.

For something like `qml.RX(x, 0) @ qml.RY(y, 0)`, we should be able to differentiate each component via parameter shift.

Comment (dwierichs) [reply pennylane/ops/op_math/composite.py:142]: Yes, for `Prod` we could have such a constructive method! It would be important to have the recipes map to the parameters of the `Prod` operator, though, and I'm not sure that's obviously guaranteed. E.g. if there is a single-parameter op with `grad_recipe=None` within the `Prod`, it's gonna shift the recipes<>parameters mapping.

Probably we could do
```python
recipes = sum(getattr(op, "grad_recipe", []) for op in self, start=[])
```
but I feel like this opens up a can of worms and is more than a bug fix...

**Comment by dwierichs** [comment]
> Definitely those big values are unexpected

I agree that they are not nice. We should fix the handling of `h` in `finite_diff`...

>On the other hand, the gradient breaks if I do things like:

Yes, this is because those examples are `SymbolicOp`s but not `CompositeOp`s, so that this fix does not apply.
I think we need to discuss the roadmap of fixing symbolic ops with `qml.gradients` more generally, this is not just a small bugfix.

File: pennylane/ops/op_math/composite.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/ops/op_math/composite.py:142]: If `grad_recipe=None`, would it have any parameters?

Comment (dwierichs) [reply pennylane/ops/op_math/composite.py:142]: Yes, it could, it's the default value for `Operation`.
